### ISSUE TITLE
FetcherThreads and DemandThreads persist after task completion

### ISSUE THREAD

**Issue by michaelhmendes** [prebid/prebid-mobile-android#156]
FetcherThreads and DemandThreads persist and new threads seem to be created for each fetch.  If the app is used long enough, there are too many threads, causing a crash.  According to our Fabric integration, the crash has the following trace:
```
Fatal Exception: java.lang.OutOfMemoryError
pthread_create (1040KB stack) failed: Try again

java.lang.Thread.nativeCreate (Thread.java)
--
  | java.lang.Thread.start + 1063 (Thread.java:1063)
  | java.util.concurrent.ThreadPoolExecutor.addWorker + 921 (ThreadPoolExecutor.java:921)
  | java.util.concurrent.ThreadPoolExecutor.processWorkerExit + 989 (ThreadPoolExecutor.java:989)
  | java.util.concurrent.ThreadPoolExecutor.runWorker + 1131 (ThreadPoolExecutor.java:1131)
  | java.util.concurrent.ThreadPoolExecutor$Worker.run + 588 (ThreadPoolExecutor.java:588)
  | java.lang.Thread.run + 818 (Thread.java:818)
```
There are additional entrypoints to this crash (org.prebid.mobile.d.<init> (d.java), org.prebid.mobile.d$a.<init> (d.java).  

According to Fabric, it affects thousands of our users every month.

**Steps to inspect the behavior:**
1. In Android Studio, navigate an app that loads Prebid ads in an emulator in Debug mode.  
2. Pause the app through the debugger, and look over the number of active threads labeled "FetcherThread" or "DemandThread"
3. Continue navigating the app and loading Prebid ads for a while, then repeat steps 1-2.  There will be more FetcherThreads and DemandThreads.
4. If this is an app where the Prebid loads occur on a secondary activity, navigate away from that activity so it's destroyed, then repeat steps 1-2.  The threads have not been terminated.

**Expected behavior:**
FetcherThread and DemandThread should terminate when done.

**Affected Users:**
 - Many devices in Production, predominantly (but not limited to) Samsung devices running Android 7.x

**Additional context**
I've attached a couple thread-trace exports, one from early on in our app's run and one from later, after many Prebid ads have been loaded, then unloaded and the activity destroyed.  The threads are still hanging around.

**Comment by michaelhmendes** [comment]
[threads_report_early.txt](https://github.com/prebid/prebid-mobile-android/files/3815750/threads_report_early.txt)

**Comment by michaelhmendes** [comment]
[threads_report_late.txt](https://github.com/prebid/prebid-mobile-android/files/3815751/threads_report_late.txt)

**Comment by michaelhmendes** [comment]
@anwzhang Let me know if you need more information or reports from our end, or if there's anything you'd like us to try.

**Comment by michaelhmendes** [comment]
I've attached a diff for the prebid-mobile-android code that seems to fix the issue.  With this fix, I believe existing applications using this library won't have to make any updates in their own code unless they use autoRefresh, in which case they need to make sure to stopAutoRefresh on the adUnits when destroying an ad.  

[issue_156_end_fetcher_demand_threads.txt](https://github.com/prebid/prebid-mobile-android/files/4054935/issue_156_end_fetcher_demand_threads.txt)

**Comment by michaelhmendes** [comment]
PR created from forked repo: https://github.com/prebid/prebid-mobile-android/pull/166

**Comment by anwzhang** [comment]
Resolved by #166

**Comment by michaelhmendes** [comment]
@anwzhang Are there plans to cut a new release of Liberty with this fix?  

If the release is coming soon, we'll wait and use that, but if there are other features planned before the next release, we can create a custom version to use locally in the meantime.
